docs: add "random fact generator" custom step workshop to tutorials (#2694)
slackapi/bolt-js: docs/english/tutorials/random-fact-generator/random-fact-generator.md

### Install app in workspace

Still in the app settings, navigate to the **Install App** page in the left sidebar. Install your app. When you press **Allow**, this means you’re agreeing to install your app with the permissions that it’s requesting. Copy the bot token that you receive and store it in a safe place for subsequent steps.

Now that you’ve completed the app setup, it’s time to write some code!

### Install Node and set up the project

Open your terminal or command prompt. First, check to see that you have Node.js installed and it is a recent long-term support (LTS) version by typing the following command.

```sh
node -v
```

If you get an error or if it’s an older version than what’s available for download on the [Node.js website](https://nodejs.org/), take a moment to install the latest version.

Next, create a new directory for the project named `useless-fact-app` and change to the new directory.

```sh
mkdir useless-fact-app
cd useless-fact-app
```

For Windows, the command looks like this:

```powershell
md useless-fact-app
```

Initialize a new Node.js project using the following command; answer each question with defaults.

```sh

> npm init

Reviewer: 📎 thought: In a follow up I'm wondering if we can start from the [blank](http://github.com/slack-samples/bolt-js-blank-template) template to remove configuration to the `package.json` file?

Author: Good callout. I'll create a follow-up task to rework this using the blank template!

```

Our project is going to use two packages: the Slack [Bolt for JavaScript](/tools/bolt-js/) framework and [Axios](https://www.npmjs.com/package/axios), a popular HTTP client for making API calls. Install these dependencies with the following command.

```sh
npm install @slack/bolt axios
```

Open the project in VS Code or your favorite code editor, then create a new file in the project folder named `.env`. Open the file and paste the following, replacing the placeholders with the values you saved earlier. 

```sh
SLACK_SIGNING_SECRET={your-app-signing-secret}
SLACK_BOT_TOKEN={your-bot-token-xoxb-1234}
SLACK_APP_TOKEN={your-app-token-xapp-1234}
PORT=3000

> ```

Reviewer: ````suggestion
```txt
SLACK_BOT_TOKEN={your-bot-token-xoxb-1234}
SLACK_APP_TOKEN={your-app-token-xapp-1234}
```
````

💸 suggestion: Related to signing secrets and socket mode!

Next, edit the `package.json` file with the following settings, then save the file.
* Change `"main"` to:

```json
"main": "app.js",
```

* Change `"scripts"` to:

```json
"scripts": {

>   "dev": "node –env-file=.env –watch app.js"

Reviewer: ```suggestion
  "dev": "node --env-file=.env --watch app.js"
```

🪬 issue: This might be formatting error?

}
```

* Change `"type"` to `"module"` or add this line if it is not already present:

```json
"type": "module"
```

### Add function code

Create a new file named `app.js` and add the following code.

```js
import bolt from "@slack/bolt";
import axios from "axios";

const { App } = bolt;

// Initialize the Bolt app
const app = new App({
  token: process.env.SLACK_BOT_TOKEN,

>   signingSecret: process.env.SLACK_SIGNING_SECRET,

Reviewer: ```suggestion
```

⚡ suggestion: With socket mode as `true` we don't need the signing secret in most cases! OAuth might be an exception?

  appToken: process.env.SLACK_APP_TOKEN,
  socketMode: true,
});

// Make an API call to retrieve a random fact
async function getUselessFact() {
  const res = await axios.get("https://uselessfacts.jsph.pl/api/v2/facts/random");
  return res.status === 200 ? res.data : null;
}

// Define the function for Workflow Builder
app.function("useless_fact_step", async ({ complete, fail, logger }) => {
  try {
    app.logger.info("Running useless fact step...");

    // Get the fact using the API
    const fact = await getUselessFact();

    if (fact && fact.text) {
      // Use complete() to send results to Slack
      await complete({
        outputs: {
          message: fact.text,
        },
      });
    } else {
      // Use fail() to send an error to Slack
      await fail({ error: "There was an error retrieving a random useless fact :cry:" });
    }
  } catch (error) {
    // Log the error and send an error message back to Slack
    logger.error(error);
    await fail({ error: `There was an error 😥 : ${error}` })
    }
});

// Start the Bolt App!
async function main() {

>   await app.start(process.env.PORT || 3000);

Reviewer: ```suggestion
  await app.start();
```

🔌 suggestion: Related to socket mode also not needing a specific port!

  app.logger.info("⚡️ Bolt app is running!");
}

main();
```

> The code in the `app.js` file initializes the Bolt app using your tokens and signing secret and enables Socket Mode. Next, it defines the `getUselessFact()` function that retrieves a random fact using Axios. It uses the Bolt framework to create a Slack function named `useless_fact_step`. This is the same function we registered in the app manifest previously; the function's name, inputs, and outputs in the code must match with the what is defined in the app manifest. Finally, a function named `main()` is used to start the Bolt app.

Reviewer: ```suggestion
The code in the `app.js` file initializes the Bolt app using your tokens and signing secret and enables Socket Mode. Next, it defines the `getUselessFact()` function that fetches a random fact. It uses the Bolt framework to create a Slack function named `useless_fact_step`. This is the same function we registered in the app manifest previously; the function's name, inputs, and outputs in the code must match with the what is defined in the app manifest. Finally, a function named `main()` is used to start the Bolt app.
```

👁️‍🗨️ suggestion: We're making "fetch" happen

5. Click **{} Insert a variable** and then select **Fact** under **1. Useless Fact Custom Step**.
6. Click **Save**.

### Publish and test the workflow

Almost there! The final task is to publish the workflow and wait for the next scheduled run to see it in action!

1. Click **Finish Up**.
2. Enter a **Name**, such as `Daily Useless Fact`.
3. Update the **Description**.
4. Click **Publish**.

![Image of finish screen](random-fact-app-2.png)

Wait for the scheduled time and see your new custom Bolt app in action!

### Make changes to the workflow

If needed, you can always make changes to elements of your workflow, such as the schedule, channel, or message. 

1. From the left sidebar in Slack, click **More**, then **Tools**, then **Workflows**. 
2. Find your workflow and click the three vertical dots on the right side.
3. Click **Edit workflow** and make changes. Don't forget to publish!

## Deploy your Bolt app

Congratulations! You’ve learned how to create a Bolt app using the [Bolt for JavaScript](/tools/bolt-js) framework and a daily scheduled workflow using [Workflow Builder](/workflows/workflow-builder)! 

> The Bolt app will continue to work as long as you have it running on your computer. This is acceptable for development and testing purposes, but a better long-term solution is to deploy the application to a production environment. You can follow the guides available for [Deploying to AWS Lambda](/tools/bolt-js/deployments/aws-lambda/) or [Deploying to Heroku](https://tools.slack.dev/bolt-js/deployments/heroku) for steps on deploying to those environments. Alternatively, you can deploy to any platform that supports JavaScript serverless functions or Node.js applications.

Reviewer: ```suggestion
The Bolt app will continue to work as long as you have it running on your computer. This is acceptable for development and testing purposes, but a better long-term solution is to deploy the application to a production environment. You can follow the guides available for [Deploying to AWS Lambda](/tools/bolt-js/deployments/aws-lambda/) or [Deploying to Heroku](/tools/bolt-js/deployments/heroku) for steps on deploying to those environments. Alternatively, you can deploy to any platform that supports JavaScript serverless functions or Node.js applications.
```

📠 todo(non-blocking): A quick redirect!
